### PR TITLE
SF.11, Header files should be self-contained

### PR DESCRIPTION
**Pull request by ChrisGuzak**
per issue #1168

### CODE REVIEW

**Comment by MikeGitb** [comment]
I think, self-contained is a more commonly used term. And maybe the some explanation would be good (e.g. just use the part from enforcement section)

**Comment by ChrisGuzak** [comment]
@MikeGitb, @hsutter suggested this title, so I'd like to hear what he thinks.

**Comment by hsutter** [comment]
Either term works for me.

File: CppCoreGuidelines.md [hunk 1]
```diff
@@ -18614,6 +18615,27 @@ This rule against implicit inclusion is not meant to prevent such deliberate agg
 Enforcement would require some knowledge about what in a header is meant to be "exported" to users and what is there to enable implementation.
 No really good solution is possible until we have modules.
 
+### <a name="Rs-sufficient"></a>SF.11: Header files should be self-sufficient
+
+##### Reason
+
+Usability, headers should be simple to use.
+Headers should encpsulate the functionality they provide.
+Avoid clients of a header having to manage its dependencies.
```
Comment (blakehawkins) [inline CppCoreGuidelines.md]: maybe "that header's" instead of "its"

**Comment by beinhaerter** [comment]
> "Usability, headers should be simple to use and work when included by themselves."

This first made be wonder why headers might include themselves. I am not a native English speaker but does the following make more sense?

"Usability. Headers should be simple to use and work by themselves when included."

It think it is also worth mentioning that if you add a dependency in a header file (so you change an existing file) but forget to add the corresponding #include then existing code might break. It might be added to the Note.

What I like to do to enforce this rule is that in file.cpp I include file.h at the top of the file to see if it compiles. But I also heard different opinions on the include order. So I think this should not be a guideline. But it might be added as a note.

**Comment by jwakely** [comment]
> does the following make more sense?

No, not really. How about "simple to use and work when included on their own" or "... when included without any other headers".

**Comment by ChrisGuzak** [comment]
I went with @jwakely's suggestion

**Comment by beinhaerter** [comment]
The enforcement section still mentions, that the header file shall include itself.

> A test should verify that each header file can be included by itself and successfully compile.

I suggest changing this. Maybe to

> A test should verify that the header file itself compiles or that a cpp file which only includes the header file compiles.

**Comment by ChrisGuzak** [comment]
I took @beinhaerter comment.

**Comment by hsutter** [comment]
Looks good, thanks Chris.

**Pull request merged**